R/replace.R

```diff
@@ -179,18 +179,50 @@ str_replace_na <- function(string, replacement = "NA") {
 
 str_transform <- function(string, pattern, replacement) {
   loc <- str_locate(string, pattern)
-  str_sub(string, loc, omit_na = TRUE) <- replacement(str_sub(string, loc))
+  new <- replacement(str_sub(string, loc))
+  str_sub(string, loc, omit_na = TRUE) <- new
   string
 }
-str_transform_all <- function(string, pattern, replacement) {
+
+str_transform_all <- function(string, pattern, replacement, error_call = caller_env()) {
   locs <- str_locate_all(string, pattern)
 
-  for (i in seq_along(string)) {
-    for (j in rev(seq_len(nrow(locs[[i]])))) {
-      loc <- locs[[i]]
-      str_sub(string[[i]], loc[j, 1], loc[j, 2]) <- replacement(str_sub(string[[i]], loc[j, 1], loc[j, 2]))
-    }
+  old <- str_sub_all(string, locs)
+  idx <- chop_index(old)
+
+  # unchop list into a vector, apply replacement(), and then rechop back into
+  # a list
+  old_flat <- vctrs::vec_unchop(old, idx)
+  if (length(old_flat) == 0) {
+    # minor optimisation to avoid problems with the many replacement
+    # functions that use paste
+    new_flat <- character()
```

Comment:
💀

Reply:
Alternate idea is to actually apply `replacement()` and then check if `length(new_flat) == 1 && length(old_flat) == 0`. In that special case you could issue a warning (possibly even mentioning `paste0()`) and "fix" it for them.

Reply:
Yeah, we could do that, I just don't think it's worth the effort since we do know what the correct answer is in the 0 element case.